lit_gpt/utils.py

```diff
@@ -24,6 +25,61 @@ def find_multiple(n: int, k: int) -> int:
     return n + k - (n % k)
 
 
+def get_batch(
+    fabric: L.Fabric,
+    data: List[Dict],
+    longest_seq_length: int,
+    micro_batch_size: int,
+    longest_seq_ix: Optional[int] = None
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    ix = torch.randint(len(data), (micro_batch_size,))
+    if longest_seq_ix is not None:
+        # force the longest sample at the beginning so potential OOMs happen right away
+        ix[0] = longest_seq_ix
+
+    input_ids = [data[i]["input_ids"].type(torch.int64) for i in ix]
+    labels = [data[i]["labels"].type(torch.int64) for i in ix]
+
+    # it's better to pad to a fixed seq length with XLA to avoid recompilation
+    max_len = max(len(s) for s in input_ids) if fabric.device.type != "xla" else longest_seq_length
+
+    def pad_right(x, pad_id):
+        # pad right based on the longest sequence
+        n = max_len - len(x)
+        return torch.cat((x, torch.full((n,), pad_id, dtype=x.dtype)))
+
+    x = torch.stack([pad_right(x, pad_id=0) for x in input_ids])
+    y = torch.stack([pad_right(x, pad_id=-1) for x in labels])
+
+    if fabric.device.type == "cuda" and x.device.type == "cpu":
+        x, y = fabric.to_device((x.pin_memory(), y.pin_memory()))
+    else:
+        x, y = fabric.to_device((x, y))
+    return x, y
+
+
+@torch.no_grad()
+def check_compatible_tokenization(
```

Comment:
There is no need to write a decoration like this. `get_batch` and the check are independent things. 

How about this:

```py
def _check_compatible_tokenization(ids, vocab_size):
 
    if (ids > ).any():
        raise ValueError(
            "Some input IDs exceed the vocabulary size of this model. "
            "Please make sure you used this model's "
            "tokenizer to prepare the dataset."
    )
```

Simply calling this in a one-liner inside get_batch would be sufficient.

Reply:
You are 100% right. That was dumb, have been staring at screens too long this week 🤦‍♂️

Reply:
ah wait, the problem is that we don't want to have this in the `validate` function, because this function gets also called later, and that would then be not ideal performance-wise. I think the idea was decoupling it from `validate`